### PR TITLE
Optimize tail-recursive functions with a base case to loops

### PR DESCRIPTION
**Pull request by oflatt**

### CODE REVIEW

File: tests/snapshots/files__block-diamond-optimize.snap [hunk 1]
```diff
@@ -8,22 +8,31 @@ expression: visualization.result
   c2_: int = const 1;
   c3_: int = const 2;
   v4_: bool = lt v0 c3_;
-  c5_: int = const 4;
-  v6_: int = select v4_ c5_ c2_;
-  v7_: int = id v6_;
+  c5_: int = const 0;
```
Comment (yihozhang) [inline tests/snapshots/files__block-diamond-optimize.snap:11]: this seems to be a small regression

File: dag_in_context/src/optimizations/swap_if.egg [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+
+;; for if statements with two outputs, swaps the order
+;; of the outputs
+(rule
+  ((= lhs (If pred inputs then else))
+   (= (tuple-length then) 2)
+   (= (tuple-length else) 2))
+  (
+    (union
```
Comment (yihozhang) [inline dag_in_context/src/optimizations/swap_if.egg:21]: Would it be more useful to union `lhs` with `(Concat (Single (Get new-lhs 1)) (Single (Get new-lhs 0)))` where `new-lhs` is the if expression with output swapped? After all, the expression being unioned here, `(Concat (Single (Get lhs 1)) (Single (Get lhs 0)))`, isn't part of the final program.

File: dag_in_context/src/schedule.rs [hunk 3]
```diff
@@ -91,6 +91,22 @@ pub fn mk_sequential_schedule() -> Vec<String> {
...
+      {helpers}
+      swap-if)
+  {helpers}
+  rec-to-loop
+  {helpers})"
+    ));
+    res.push(format!(
+        "
+;; HACK: when INLINE appears in this string
+;; we perform inlining in this pass
```
Comment (yihozhang) [inline dag_in_context/src/schedule.rs:107]: To make this comment clear, "when INLINE appears in this string, we perform inlining in this pass" => "when the keyword INLINE appears in the schedule, `optimize` in `lib.rs` inlines the input program before running this schedule"

Alternatively, we can have this function returns a `Schedule` struct, which makes this no longer a hack.

**Review by yihozhang — APPROVED** [review]

File: dag_in_context/src/optimizations/swap_if.egg (hunk 2 above)
Comment (oflatt) [reply dag_in_context/src/optimizations/swap_if.egg:21]: I believe that our tuple helpers, which create gets for all of the tuples, do discover that equivalence

File: dag_in_context/src/schedule.rs (hunk 3 above)
Comment (oflatt) [reply dag_in_context/src/schedule.rs:107]: Good idea- will follow up with another PR

**Comment by oflatt** [comment]
Merging to unblock you @yihozhang

File: tests/snapshots/files__block-diamond-optimize.snap (hunk 1 above)
Comment (oflatt) [reply tests/snapshots/files__block-diamond-optimize.snap:11]: will look into it

**Pull request merged**